### DISCUSSION TITLE
Implementing a constraint inside $def using the field value from a higher level schema

### DISCUSSION THREAD

**Discussion by girish-tharwani** [json-schema-org/community#598] [Q&A] [answered]
Hello Experts
Below is a simplified / sanitized view of my actual schema. Is it possible to implement a constraint on fieldD (lets say a constraint on array size) using the value of fieldA? e.g. (if fieldA = true then fieldD minItem = 1)

```json
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "type": "object",
    "properties": {
        "fieldA": {
            "type": "boolean"
        },
        "fieldB": {
            "type": "object",
            "$ref": "#/$defs/fieldB"
        }
    },
    "$defs": {
        "fieldB": {
            "type": "object",
            "properties": {
                "fieldC": {
                    "type": "string"
                },
                "fieldD": {
                    "$ref": "#/$defs/fieldD"
                }
            }
        },
        "fieldD": {
            "type": "array"
        }
    }
}
```

**Comment by jdesrosiers** [comment, marked as answer]
There's no way to write constraints that make use of values "above" it in the JSON tree. You have to put your `if`/`then` at a level high enough that every value involved is below the `if`/`then`.

```json
"if": {
  "properties": {
    "fieldA": { "const": true }
  },
  "required": ["fieldA"]
},
"then": {
  "properties": {
    "fieldB": {
      "properties": {
        "fieldD": { "minItems": 1 }
      }
    }
  }
}
```

Notice the nested subschemas in `then` to drill down from the top level to `fieldD`.

**Reply by girish-tharwani** [reply]
Thanks. This works perfectly. 
However I have a variation of this schema in which fieldB is an array of objects and it is defined as:
```
"fieldB": {
       "type": "array",
        "items": { "$ref" : "#/$defs/fieldB"}
      }
```
The validation doesn't work in this case. Is there a way to make it work for this case?

Edit 1
I tried using items keyword instead of properties keyword in the constraint but didn't work?

Edit 2 
Was able to make it work after a few attempts of getting the nesting of "items" and "properties" right.
